### ISSUE TITLE
Timeout while debugging

### ISSUE THREAD

**Issue by Startigers** [ignatov/intellij-erlang#696]
Hi,
when I debug a Eunit test and the debugger is at a breakpoint then the test will be canceled after about 5 seconds. 
```erlang
Test suite was cancelled. Reason: undefinedTest suite was cancelled. Reason: {blame,[1,1]}
```
This happens when I have more than one test function, for instance:
```erlang
-module(small_test).
-include_lib("eunit/include/eunit.hrl").

simple_test() ->
    ?assert(true).   %% Breakpoint 
simple2_test() ->
    ?assert(true).
```

**Comment by deadok22** [comment]
The plugin simply runs this in your case:
```erlang
 eunit:test({timeout, 16#7fffffff, [small_test]})
```

Unfortunately, ```eunit``` treats the passed timeout value as a timeout for a whole run, and not for each test in a module. Could you dig in and find if there's a way to run it properly?

You can call ```timer:sleep/1``` in your test to emulate hitting a breakpoint.

**Comment by Startigers** [comment]
I found the reason in a thread of an erlang mailing list:

_Yes, the {timeout,1,...} applies to the whole subset of tests, in this case the setup and all the tests in it. The individual tests still get their default timeout of 5 seconds. If you reduce the timeout for the whole group to 1 second, then that will of course happen first and abort all the tests in the group, while in the first case it's one of the individual tests that times out first. Only when applied to a single test does {timeout,T,...} change the timeout of the test itself._

_I'm sorry that this is confusing. There should probably be a separate way of modifying the default timeout for all individual tests._
(http://osdir.com/ml/erlang-programming-bugs/2012-02/msg00006.html)


According to this I had to use test generators. There I can set the timeout individual for every test case.
(Is there another way to set the timeout?)
```Erlang
simple1_test_() ->
    {timeout,16#7fffffff,[?_assert(simple_test1())]}.

simple_test1() ->
    timer:sleep(10000),  %% Simulated Breakpoint
    ?assert(true),
    true.

simple2_test_() ->
    {timeout,16#7fffffff,[?_assert(simple_test2())]}.

simple_test2() ->
    ?assert(true),
    true.
```
Or to run it parallel:
```Erlang
simple_test_() ->
    {inparallel, [
        {timeout,16#7fffffff,[?_assert(simple_test1())]},
        {timeout,16#7fffffff,[?_assert(simple_test2())]}
        ]
    }.

simple_test1() ->
    timer:sleep(10000),    %% Simulated Breakpoint
    ?assert(true),
    true.

simple_test2() ->
    ?assert(true),
    true.
```

In summary: It isn't a problem of the IDEA Erlang plugin. It is only a problem of the limited configuration possibilities of Eunit. How could the plugin help to avoid this problem?

**Comment by deadok22** [comment]
There are several things we can do.

First, if a tests run fails on timeout while in debug mode, we can show a notification explaining why it might have happened.

Second, we can aid users converting their tests to test generator form, and setting test timeout by adding intentions.

Yet it's very unfortunate we can't set timeouts for individual tests from the plugin.

What do you think?

**Comment by Startigers** [comment]
Yes I think this actions would assist the user in writing debuggable tests. But I have to say that I am not an Erlang expert. I use it only for a few months and not all the time cause mainly I develop in Java. 
So perhaps there is a better solution than the test generators?

**Comment by deadok22** [comment]
I believe, there isn't.

As you've already said, if you run a single test, it works fine (because the timeout is set by the plugin). Why do you want to debug a whole test suite?

**Comment by Startigers** [comment]
That is a good question. I think I debugged it only cause I executed all the tests and saw an exception. I used the same run configuration for debugging. I didn't knew that this can cause a problem.

When I debug a single test then debugging is not a problem. But the test times out when it takes longer than the default timeout (5s). See following example:

```Erlang
-module(small5_test).
-include_lib("eunit/include/eunit.hrl").

simple_test() ->
    timer:sleep(10000),
    ?assert(true).
```
So your argument that the timeout is set correctly by the plugin is not correct:
```Erlang
/opt/erlang/17.5/bin/erl -pa /tmp/eunit_teamcity93.tmp -pa /home/startigers/workspace/smalltest/target/test-classes -pa /home/startigers/workspace/smalltest -eval "eunit:test([small5_test], [{report, {eunit_teamcity,[]}}, {no_tty, true}])." -s init stop -noshell {no_tty, true}])." -s init stop -noshell
Testing started at 09:20 ...
Test suite was cancelled. Reason: {blame,[1]}
```

In the erl shell it works:
```Erlang
1> eunit:test({test,small5_test,simple_test}, [verbose]). %% default timeout
======================== EUnit ========================
small5_test: simple_test...*timed out*
undefined
=======================================================
  Failed: 0.  Skipped: 0.  Passed: 0.
One or more tests were cancelled.
error
2> eunit:test({timeout, 30 ,{test,small5_test,simple_test}}, [verbose]).
======================== EUnit ========================
small5_test: simple_test...[10.006 s] ok
=======================================================
  Test passed.
ok
3> eunit:test({timeout, 30 ,[small5_test]}, [verbose]).                 
======================== EUnit ========================
small5_test: simple_test (module 'small5_test')...[10.003 s] ok
=======================================================
  Test passed.
ok
```

**Comment by deadok22** [comment]
That's exactly what I said - you can debug a single test and in that case the plugin sets the timeout. 
In case you're not debugging your test, there is no point in altering tests timeout. We can add timeout settings to eunit run configurations, though. But I don't think it's that common to set timeouts for your tests externally, or is it?

**Comment by Startigers** [comment]
Okay, now I am on your side. Normally a unit test shouldn't take longer than the default timeout. 
And if I ever need longer timeouts I will set it using test generators.

Thanks a lot for your help.

**Comment by deadok22** [comment]
You're welcome =)
